src/ai/index.js

```diff
@@ -0,0 +1,550 @@
+        let bestWorldObject = realityEditor.worldObjects.getBestWorldObject();
+        let frames = [];
+        for (let frame of Object.values(bestWorldObject.frames)) {
+            if (!result.includes(frame.src)) continue;
+            frames.push(frame);
+        }
+        pushToolDialogue(frames, result);
+    }
+    
+    function showDialogue() {
+        aiContainer.style.animation = `slideToRight 0.2s ease-in forwards`;
+    }
+    
+    function hideDialogue() {
+        aiContainer.style.animation = `slideToLeft 0.2s ease-in forwards`;
+    }
+    
+    function hideEndpointApiKeyAndShowSearchTextArea() {
+        endpointArea.style.display = 'none';
+        apiKeyArea.style.display = 'none';
+        searchTextArea.style.display = 'block';
+        adjustTextAreaSize();
+    }
+
+    function setupEventListeners() {
+        endpointArea.addEventListener('keydown', (e) => {
+            e.stopPropagation();
+            if (e.key === 'Enter') {
+                e.preventDefault();
+                if (endpointArea.value === '' || apiKeyArea.value === '') return;
+                realityEditor.network.postAiApiKeys(endpointArea.value, apiKeyArea.value, true);
+            }
+        })
+        apiKeyArea.addEventListener('keydown', (e) => {
+            e.stopPropagation();
+            if (e.key === 'Enter') {
+                e.preventDefault();
+                if (endpointArea.value === '' || apiKeyArea.value === '') return;
+                realityEditor.network.postAiApiKeys(endpointArea.value, apiKeyArea.value, true);
+            }
+        })
+        
+        searchTextArea.addEventListener('input', function() {
+            // adjustTextAreaSize();
+        });
+        
+        // todo Steve: below 4 doesn't work
+        searchTextArea.addEventListener('mousedown', (e) => {e.stopPropagation();});
+        searchTextArea.addEventListener('mouseup', (e) => {e.stopPropagation();});
+        searchTextArea.addEventListener('mousemove', (e) => {e.stopPropagation();});
+        searchTextArea.addEventListener('contextmenu', (e) => {
+            e.stopPropagation();
+        })
+        
+        document.addEventListener('keydown', (e) => {
+            if (e.key === 'g' || e.key === 'G') {
+                // console.log('%c G clicked', 'color: red');
+            }
+        })
+
+        searchTextArea.addEventListener('focus', () => {
+            // console.log('focus');
+            // realityEditor.device.keyboardEvents.openKeyboard();
+        });
+
+        searchTextArea.addEventListener('blur', () => {
+            // console.log('blur');
+            // realityEditor.device.keyboardEvents.closeKeyboard();
+        });
+
+        searchTextArea.addEventListener('keydown', (e) => {
+            // todo Steve: bug: press g/G still focuses the camera to target. Is it b/c remote operator event listener is added later than searchTextBar listener? Figure out how the orders & layers work
```

Comment:
If you want to test whether it's the order, you can pass `true` as the last argument to the key* event listeners to capture the event earlier in the phase. However, the actual bug causing this problem is that the g/G listener is using 'keypress' and here you look for keydown and keyup

Reply:
Right. I'll add a proper focus event in the remote operator to make sure it's nicely integrated in the app, instead of making a standalone keypress listener.